cookbook/retrieval_in_sql.ipynb

```diff
@@ -0,0 +1,693 @@
+    "'SELECT \"Track\".\"Name\" FROM \"Track\" ORDER BY \"embeddings\" <-> '[loneliness]' LIMIT 5'\n",
+    "\n",
+    "If you need to combine embeddings from different tables, you can use the WITH statement:\n",
+    "WITH table1 AS (\n",
+    "    SELECT \"ColumnName\"\n",
+    "    FROM \"TableName\"\n",
+    "    ORDER BY \"embeddings\" <-> '[keyword_1]'\n",
+    "    LIMIT 5\n",
+    ")\n",
+    "SELECT \"table2\".\"Column_Name\", Table1.\"TableName\".\"ColumnName\"\n",
+    "FROM \"table2\n",
+    "JOIN table1 ON \"table2\".\"Column_Name\" = table1.\"ColumnName\"\n",
+    "ORDER BY \"table2\".\"embeddings\" <-> '[keyword_2]'\n",
+    "LIMIT 3;\n",
+    "\n",
+    "Use the following format:\n",
+    "\n",
+    "Question: Question here\n",
+    "SQLQuery: SQL Query to run\n",
+    "SQLResult: Result of the SQLQuery\n",
+    "Answer: Final answer here\n",
+    "\n",
+    "Only use the following tables:\n",
+    "{table_info}\n",
+    "\n",
+    "Question: {input}\n",
+    "\"\"\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 120,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "db = SQLDatabase.from_uri(CONNECTION_STRING) # We reconnect to dbso the new columns are loaded as well.\n",
+    "llm = ChatOpenAI(model_name='gpt-4', temperature=0)\n",
+    "chain = create_sql_query_chain(llm, db, prompt=prompt)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Using the Chain"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Example 1: Filtering a column based on semantic meaning"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's say we want to retrieve songs that express sadness, but filtering based on genre:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 30,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "SELECT \"Track\".\"Name\" \n",
+      "FROM \"Track\" \n",
+      "JOIN \"Genre\" ON \"Track\".\"GenreId\" = \"Genre\".\"GenreId\" \n",
+      "WHERE \"Genre\".\"Name\" = 'Rock' \n",
+      "ORDER BY \"Track\".\"embeddings\" <-> '[dispair]' \n",
+      "LIMIT 5\n"
+     ]
+    }
+   ],
+   "source": [
+    "question = \"Which are the 5 rock songs with titles about deep feeling of dispair?\"\n",
+    "query = chain.invoke({\"question\": question})\n",
+    "print(query)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As we can see, we have a placeholder provided by the LLM for us to insert the actual embedding for `dispair`. \n",
+    "\n",
+    "Let's use this function to replace the embedding placeholder with the actual embedding:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 31,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "\n",
+    "def replace_brackets(match):\n",
```

Comment:
@manuel-soria we can add this as a function call at the end of the LCEL chain.

We can pipe it to a function, as done here:

```
# Prompt
from langchain.memory import ConversationBufferMemory
from langchain.prompts import ChatPromptTemplate, MessagesPlaceholder
template = """Based on the table schema below, write a SQL query that would answer the user's question:
{schema}

Question: {question}
SQL Query:"""
prompt = ChatPromptTemplate.from_messages([
    ("system", "Given an input question, convert it to a SQL query. No pre-amble."),
    MessagesPlaceholder(variable_name="history"),
    ("human", template)
])

memory = ConversationBufferMemory(return_messages=True)

# Chain to query with memory 
from langchain.schema.runnable import RunnableLambda

sql_chain = (
    RunnablePassthrough.assign(
       schema=get_schema,
       history=RunnableLambda(lambda x: memory.load_memory_variables(x)["history"])
    )| prompt
    | llm.bind(stop=["\nSQLResult:"])
    | StrOutputParser()
)

def save(input_output):
    output = {"output": input_output.pop("output")}
    memory.save_context(input_output, output)
    return output['output']
    
sql_response_memory = RunnablePassthrough.assign(output=sql_chain) | save
sql_response_memory.invoke({"question": "What team is Klay Thompson on?"})
```

https://github.com/langchain-ai/langchain/blob/master/cookbook/LLaMA2_sql_chat.ipynb